src/detect.rs

```diff
@@ -18,3 +23,221 @@ pub enum DetectOutcome {
     Pass(BuildPlan),
     Fail,
 }
+
+pub struct DetectContextBuilder<P: Platform, BM: DeserializeOwned> {
+    app_dir: Option<PathBuf>,
+    buildpack_dir: Option<PathBuf>,
+    stack_id: String,
+    platform: Option<P>,
+    platform_dir: Option<PathBuf>,
+    buildpack_toml: Option<BuildpackToml<BM>>,
+    buildpack_toml_path: Option<PathBuf>,
+    buildpack_toml_string: Option<String>,
+}
+impl<P: Platform, BM: DeserializeOwned> DetectContextBuilder<P, BM> {
+    pub fn new(stack_id: impl AsRef<str>) -> Self {
+        DetectContextBuilder {
+            app_dir: None,
+            buildpack_dir: None,
+            buildpack_toml_path: None,
+            buildpack_toml_string: None,
+            platform_dir: None,
+            stack_id: String::from(stack_id.as_ref()),
+            platform: None,
```

Comment:
Should we default this to `GenericPlatform` if none is defined?

Reply:
We need to be lazy about it and set it in the `build()` otherwise we lose the programmer intent. I'm fine with defaulting more things. We could get rid of this error and replace it with a GenericPlatform instead.

```
                    return Err(crate::error::Error::Other(
                        "Platform is missing, call platform() or platform_from_path()".to_string(),
                    ))
```

It's ultimately a design question if that would be "unexpected" and we want the programmer to be more explicit/intentional or we want it to JUST WORK ™️. 

I guess as I'm typing this out, I'm realizing that if they weren't expecting a GenericPlatform, and got one then their types would likely get borked and they would find out that way. In that case...I'm in favor of handling stuff for the user where possible.

Reply:
Actually, scratch that, it looks like we need a dir to build one `GenericPlatform::from_path(&platform_dir)`